vowpalwabbit/slim/test/vw_util.cpp

```diff
@@ -0,0 +1,100 @@
+#include "vw_util.h" 
+#include "hash.h"
+#include "constant.h"
+
+#ifdef BOOST_FOUND
+
+#include <boost/spirit/include/qi.hpp>
+#include <boost/spirit/include/phoenix_core.hpp>
+#include <boost/spirit/include/phoenix_operator.hpp>
+#include <boost/spirit/include/phoenix_stl.hpp>
+#include <boost/bind.hpp>
+
+#endif 
+
+using namespace vw_slim;
+
+#ifdef BOOST_FOUND
+
+struct example_features_push_back_impl
+{
+  struct result { typedef void type; };
+
+  void operator()(safe_example_predict& c, namespace_index const& ns, uint64_t namespace_hash, feature_index const& idx, feature_value const& x) const
+  { c.feature_space[ns].push_back(x, namespace_hash + idx);  }
+};
+
+boost::phoenix::function<example_features_push_back_impl> const example_features_push_back = example_features_push_back_impl();
+
+struct vw_hash_impl
+{
+	struct result { typedef uint64_t type; };
+
+	uint64_t operator()(std::vector<char>& s, uint64_t seed = 0) const
+	{ return uniform_hash((const void*)&s[0],s.size(), seed); }
+};
+
+boost::phoenix::function<vw_hash_impl> const vw_hash = vw_hash_impl();
+
+struct first_char_impl
+{
+	struct result { typedef char type; };
+
+	char operator()(std::vector<char>& s) const
+	{ return s[0]; }
+};
+
+boost::phoenix::function<first_char_impl> const first_char = first_char_impl();
+
+template <typename Iterator>
+bool try_parse_examples(Iterator first, Iterator last, safe_example_predict& ex)
+{
+  namespace ascii = boost::spirit::ascii;
```

Comment:
What are spirit and phoenix and do we need them?

Reply:
fast parser for floats.

Reply:
VW contains a float parser already, can that be reused?

Reply:
I responded in a different comment about the boost Spirit.  I am not aware of the exact reason for implementing a new parser for vw text format here.  I can think of a few reasons.

1) vw_core dependencies might have to be minimized.  
2) A verifiably correct parser might be needed.  Since this parser specifies the exact grammar, valid and invalid input can be verified by inspection.  This is difficult with a state machine based parsing without documenting the actual state machine

I don't know how performant this parser is compared to the hand coded one in vw.  This should be compared.  If this is found to be a better or equivalent I would recommend either this or another parser generator for vw text format.  I have used the gold parser generator which is a table driven parser with good performance characteristics.

This does not seem to be used in rest of vw_slim.  Likely used in a utility by 3rdparty.  Boost is included only if found.  I am inclined to leave this as is till this code base is consumed from here by contributors to minimize incompatibilities.

---
In reply to: [304042853](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1800#discussion_r304042853) [](ancestors = 304042853)